manticore/core/cpu/disasm.py

```diff
@@ -0,0 +1,281 @@
+        # FIXME how to deal with discrepancies of binja vs real program
+        # entry point addresses? We need to lookup the symbols and
+        # we should make sure all offsets are appropriate
+
+        if self.entry_point_diff is None:
+            # assume that the first time we are called, this is the entry point
+            # self.entry_point_diff = addr - self.view.entry_point
+            self.entry_point_diff = 0
+
+        return addr - self.entry_point_diff
+
+    def _pop_from_il_queue(self, code, pc):
+        # queue contains tuples of the form (idx, il_insn)
+        if self.il_queue != []:
+            # if we have multiple ils for the same pc
+            # pop from the front of the queue
+            if self.il_queue[0][1].address == pc:
+                return self.il_queue.pop(0)[1]
+            else:
+                # somehow we have a queue but we are now at a different pc,
+                # clear the queue (we might be here because of a CALL etc.)
+                # FIXME assert that this is legitimate
+                del self.il_queue[:]
+
+        from binaryninja import Architecture, LowLevelILFunction
+
+        func = LowLevelILFunction(self.view.arch)
+        func.current_address = pc
+        self.disasm_insn_size = (self.view.arch.
+                                 get_instruction_low_level_il(code, pc, func))
+        self.current_llil_func = func
+        self.il_queue = [(i, func[i]) for i in xrange(len(func))]
+        return self.il_queue.pop(0)[1]
+
+    def disassemble_instruction(self, code, pc):
+        """Get next instruction based on Capstone disassembler
+
+        :param code: disassembled code
+        :param pc: program counter
+        """
+        import binaryninja.enums as enums
+
+        pc = self._fix_addr(pc)
+
+        # FIXME will be removed
+        ##################
+        blocks = self.view.get_basic_blocks_at(pc)
+        if not blocks:
+            # Looks like Binja did not know about this PC..
+            self.view.create_user_function(pc)
+            self.view.update_analysis_and_wait()
+            self.current_func = self.view.get_function_at(pc)
+        else:
+            self.current_func = blocks[0].function
+        ##################
+
+        il = self._pop_from_il_queue(code, pc)
+        self.insn_size = il.size
+        self.current_pc = pc
+        self.disasm_il = il
+
+        # create an instruction from the fallback disassembler if Binja can't
+        o = il.operation
+        if (o == enums.LowLevelILOperation.LLIL_UNIMPL or
+                o == enums.LowLevelILOperation.LLIL_UNIMPL_MEM):
+            return self.fallback_disasm.disassemble_instruction(code, pc)
+
+        return self.BinjaILInstruction(il,
+                                       self.entry_point_diff,
+                                       self.disasm_insn_size,
+                                       self.current_llil_func)
+
+
+    class BinjaILInstruction(Instruction):
```

Comment:
The idea here is that each Disassembler will needs its own Instruction spec, thus it is logically
within that scope. No strong opinions though